### ISSUE TITLE
Race condition during installation of .keep files to the same directory

### ISSUE THREAD

**Issue by mniestroj** [netdata/netdata#7292]
##### Bug report summary
Installlation phase sometimes fails, when trying to install `.keep` files into filesystem with symlinks instead of read directories, when at least two symlink point at the same directory.

##### OS / Environment
Linux, [Buildroot](https://buildroot.org/)
https://patchwork.ozlabs.org/patch/1186612/

##### Netdata version (ouput of `netdata -V`)
v1.18.1

##### Component Name
autotools installation phase

##### Steps To Reproduce
This bug is reproducible only when multiple paths to which `.keep` files are installed resolve to single directory. In case of default [Buildroot](https://buildroot.org/) configuration this is true for `/var/log/netdata` and `/var/cache/netdata` paths. `/var/log` and `/var/cache` are symlinks to `/tmp`. During installation phase two `install` processes are spawned to put `.keep` into `/tmp/netdata`. This is when the race starts. Installation procedure fails when this happens:
1) `install(1)` check whether `/tmp/netdata/.keep` is already there and unlinks it in that case 
2) `install(1)` creates file at `/tmp/netdata/.keep`
3) `install(2)` checks whether `/tmp/netdata/.keep` is there and unlinks it
4) `install(1)` tries to change permission to `/tmp/netdata/.keep`, but it fails because there is no such file
5) `install(2)` creates a `/tmp/netdata/.keep`
6) `install(2)` changes permission to `/tmp/netdata/.keep`

At the end the directory structure is okay, but installation stops, because 4) step has failed. 
##### Expected behavior
Installation phase exists with code 0, which means success.

**Comment by mniestroj** [comment]
One of the possible ways to solve this issue is to (have option to) remove installation of `.keep` files for `/var/log/netdata` and `/var/cache/netdata` directories. Those are not even shipped within Buildroot rootfs generation, so we need to create those directories during runtime anyway. This is also the simplest solution for now, so we will most likely go for it initially in Buildroot. However, as the move to CMake is ongoing, I see no point in fixing this issue now with autotools. It would be enough to just take it into account during migration to CMake.

**Comment by amoss** [comment]
Thank you for the very detailed and well worked-out bug report. The buildroot configuration that you are maintaining looks very useful and we would like to help solve this problem with you. Although the migration to CMake has started... it's a long road :). We can take a look at the autotools config and see if we can improve this.

@knatsakis I took a look through the mailing-list discussion and https://pastebin.com/raw/P9DQC1ry is interesting. My first question would be: do we really need those `.keep` entries in the target file-system at all? I'm used to putting them in working-copies to get around the default behaviour of git with empty directories, but I don't immediately have an idea of why we need them on the target system.

**Comment by Ferroin** [comment]
It's not just git that behaves like this, some package mangers have similar behavior (for example, Portage on Gentoo does, though they also name the file such that they know which package created it (effectively reference counting otherwise empty directories)).

**Comment by amoss** [comment]
What would the effect be on our directory structure post-installation? I'm wondering if we need to be concerned that directories without a `.keep` will be:

* Missed during uninstall.
* Removed by something cleaning up the system.

If not - then we can remove the `.keep` files and solve the problem that way. But if we need to maintain them for an external reason then we will have to think of a different solution.

**Comment by knatsakis** [comment]
Hi @mniestroj, thank you very much for your bug report.
I am trying to find some time to look into this and get back to you. 

@amoss my first impression is that we don't really need the `.keep` files and we should just stop installing them, but I am not 100% sure yet.

**Comment by prologic** [comment]
Unfortuantely I was not able to reproduce the race condition as described by @mniestroj here.

This was my repro steps:

```sh
$ git clone https://github.com/prologic/netdata
$ cd netdata
cat > test.sh <<EOF
#!/bin/sh

apk add alpine-sdk bash curl libuv-dev zlib-dev util-linux-dev libmnl-dev gcc make git autoconf automake pkgconfig python logrotate

mv /var/cache/* /tmp
mv /var/log/* /tmp
rm -rf /var/cache /var/log

cd /var || exit 1
ln -s /tmp cache
ln -s /tmp log

cd /netdata
./netdata-installer.sh
```
$ docker run -i -t --rm -v $PWD:/netdata -w /netdata alpine ./test.sh
```

I _tried _ produce a script that _thought_ matched @mniestroj 's repro steps in the original description; but I cannot get the build/install to fail.

That being said the correct fix _I believe_ is in #7829 and that PR is no up for review. Basically @amoss and @knatsakis are right -- We do not need these `.keep` files anywhere, they do not need to be in our Git repo as we have no empty directories to speak of, we don't need them in our build tools or in shipped  packages (_which are usually mostly tarballs_).

So they're gone now :) -- If any issues arrises from this PR (_once merged_) I'll take full responsibility and "fix forward" :) -- I say this because its really hard to understand why we did this in the first place and whether or not there will be any unforeseen side-effects of my change/fix :)
